File: c/universalfft.c
/**
 * universalfft.c — Boteler (2012) FFT conventions, self-contained C99.
 *
 * FFT algorithm: Cooley–Tukey radix-2 decimation-in-time.
 * Bit-reversal permutation then butterfly passes.
 */

#include "universalfft.h"

#include <math.h>
#include <string.h>
#include <stdlib.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

/* -----------------------------------------------------------------------
 * Internal helpers
 * ----------------------------------------------------------------------- */

static int is_power_of_two(size_t n) {
    return n > 0 && (n & (n - 1)) == 0;
}

static void bit_reverse_permute(double *re, double *im, size_t N) {
    size_t j = 0;
    for (size_t i = 1; i < N; i++) {
        size_t bit = N >> 1;
        for (; j & bit; bit >>= 1) j ^= bit;
        j ^= bit;
        if (i < j) {
            double tmp;
            tmp = re[i]; re[i] = re[j]; re[j] = tmp;
            tmp = im[i]; im[i] = im[j]; im[j] = tmp;
        }
    }
}

/* -----------------------------------------------------------------------
 * Core: in-place Cooley–Tukey radix-2 DIT FFT.
 * sign = -1 → forward (e^{-i2πkn/N}), +1 → inverse (before 1/N).
 * ----------------------------------------------------------------------- */
int ufft_inplace(double *x_re, double *x_im, size_t N, int inverse) {
    if (!is_power_of_two(N)) return -1;

    bit_reverse_permute(x_re, x_im, N);

    double sign = inverse ? +1.0 : -1.0;

    for (size_t len = 2; len <= N; len <<= 1) {
        double ang = sign * 2.0 * M_PI / (double)len;
        double wr = cos(ang);
        double wi = sin(ang);

        for (size_t i = 0; i < N; i += len) {
            double wre = 1.0, wim = 0.0;
            for (size_t j = 0; j < len / 2; j++) {
                size_t u = i + j;
                size_t v = i + j + len / 2;
                /* butterfly: t = w * x[v] */
                double tre = wre * x_re[v] - wim * x_im[v];
                double tim = wre * x_im[v] + wim * x_re[v];
                x_re[v] = x_re[u] - tre;
                x_im[v] = x_im[u] - tim;
                x_re[u] += tre;
                x_im[u] += tim;
                /* w *= (wr + i*wi) */
                double wre_new = wre * wr - wim * wi;
                wim = wre * wi + wim * wr;
                wre = wre_new;
            }
        }
    }
    return 0;
}

/* -----------------------------------------------------------------------
 * CC forward: X[k] = Σ x[n] e^{-i2πkn/N} * dt
 * ----------------------------------------------------------------------- */
int ufft_cc_forward(
    const double *in_re, const double *in_im,
    double *out_re, double *out_im,
    size_t N, double dt)
{
    memcpy(out_re, in_re, N * sizeof(double));
    memcpy(out_im, in_im, N * sizeof(double));
    int rc = ufft_inplace(out_re, out_im, N, 0);
    if (rc != 0) return rc;
    for (size_t k = 0; k < N; k++) {
        out_re[k] *= dt;
        out_im[k] *= dt;
    }
    return 0;
}

/* -----------------------------------------------------------------------
 * CC inverse: x[n] = Σ X[k] e^{+i2πkn/N} * Δf   (Δf = 1/(N*dt))
 * NumPy-ifft applies 1/N; we need Δf = 1/(N*dt), so factor = N * Δf = 1/dt.
 * ----------------------------------------------------------------------- */
int ufft_cc_inverse(
    const double *in_re, const double *in_im,
    double *out_re, double *out_im,
    size_t N, double dt)
{
    memcpy(out_re, in_re, N * sizeof(double));
    memcpy(out_im, in_im, N * sizeof(double));
    int rc = ufft_inplace(out_re, out_im, N, 1);   /* inverse twiddles */
    if (rc != 0) return rc;
    /* ufft_inplace does NOT apply 1/N — we scale by Δf = 1/(N*dt) */
    double df = 1.0 / ((double)N * dt);
    for (size_t n = 0; n < N; n++) {
        out_re[n] *= df;
        out_im[n] *= df;
    }
    return 0;
}

/* -----------------------------------------------------------------------
 * CD forward: X[k] = (1/N) Σ x[n] e^{-i2πkn/N}
 * ----------------------------------------------------------------------- */
int ufft_cd_forward(
    const double *in_re, const double *in_im,
    double *out_re, double *out_im,
    size_t N)
{
    memcpy(out_re, in_re, N * sizeof(double));
    memcpy(out_im, in_im, N * sizeof(double));
    int rc = ufft_inplace(out_re, out_im, N, 0);
    if (rc != 0) return rc;
    double inv_N = 1.0 / (double)N;
    for (size_t k = 0; k < N; k++) {
        out_re[k] *= inv_N;
        out_im[k] *= inv_N;
    }
    return 0;
}

/* -----------------------------------------------------------------------
 * CD inverse: x[n] = Σ X[k] e^{+i2πkn/N}  (raw summation — no 1/N)
 * ----------------------------------------------------------------------- */
int ufft_cd_inverse(
    const double *in_re, const double *in_im,
    double *out_re, double *out_im,
    size_t N)
{
    memcpy(out_re, in_re, N * sizeof(double));
    memcpy(out_im, in_im, N * sizeof(double));
    /* forward twiddles with conjugate sign = inverse, no 1/N */
    return ufft_inplace(out_re, out_im, N, 1);
}

/* -----------------------------------------------------------------------
 * Frequency bin array
 * ----------------------------------------------------------------------- */
void ufft_freqs(double *f, size_t N, double dt) {
    double df = 1.0 / ((double)N * dt);
    for (size_t k = 0; k < N; k++) {
        if (k < N / 2)
            f[k] = (double)k * df;
        else
            f[k] = ((double)k - (double)N) * df;
    }
}

/* -----------------------------------------------------------------------
 * Filter via CC pair
 * ----------------------------------------------------------------------- */
int ufft_filter(
    const double *in_re, const double *in_im,
    const double *H_re,  const double *H_im,
    double *out_re, double *out_im,
    size_t N, double dt)
{
    double *X_re = malloc(N * sizeof(double));
    double *X_im = malloc(N * sizeof(double));
    if (!X_re || !X_im) { free(X_re); free(X_im); return -2; }

    int rc = ufft_cc_forward(in_re, in_im, X_re, X_im, N, dt);
    if (rc != 0) { free(X_re); free(X_im); return rc; }

    /* complex multiply: Y = X * H */
    for (size_t k = 0; k < N; k++) {
        double yr = X_re[k] * H_re[k] - X_im[k] * H_im[k];
        double yi = X_re[k] * H_im[k] + X_im[k] * H_re[k];
        X_re[k] = yr;
        X_im[k] = yi;
    }

    rc = ufft_cc_inverse(X_re, X_im, out_re, out_im, N, dt);
    free(X_re);
    free(X_im);
    return rc;
}

File: fortran/universalfft.f90
!> UniversalFFT — Boteler (2012) compliant FFT/IFFT in Fortran 90.
!>
!> Self-contained Cooley-Tukey radix-2 DIT FFT — no external dependencies.
!> N must be a power of two.
!>
!> Convention (same as C implementation):
!>   CC forward : X(k) = Σ x(n) e^{-i2πkn/N} Δt
!>   CC inverse : x(n) = Σ X(k) e^{+i2πkn/N} Δf   (Δf = 1/(N·Δt))
!>   CD forward : X(k) = (1/N) Σ x(n) e^{-i2πkn/N}
!>   CD inverse : x(n) = Σ X(k) e^{+i2πkn/N}       (raw sum)
module universalfft_mod
    use iso_fortran_env, only: dp => real64
    implicit none
    private
    public :: ufft_inplace, ufft_cc_forward, ufft_cc_inverse, &
              ufft_cd_forward, ufft_cd_inverse, ufft_freqs,   &
              ufft_filter, ufft_low_pass

    real(dp), parameter :: PI = 3.14159265358979323846_dp

contains

    ! ── Bit-reversal permutation ─────────────────────────────────────────────
    subroutine bit_reverse(xr, xi, N)
        integer,    intent(in)    :: N
        real(dp),   intent(inout) :: xr(0:N-1), xi(0:N-1)
        integer :: i, j, bit
        real(dp) :: tmp
        j = 0
        do i = 1, N-1
            bit = N / 2
            do while (iand(j, bit) /= 0)
                j = ieor(j, bit)
                bit = bit / 2
            end do
            j = ieor(j, bit)
            if (i < j) then
                tmp = xr(i); xr(i) = xr(j); xr(j) = tmp
                tmp = xi(i); xi(i) = xi(j); xi(j) = tmp
            end if
        end do
    end subroutine

    ! ── Core in-place Cooley-Tukey FFT ───────────────────────────────────────
    !> inverse=0 → forward (e^{-i2πkn/N}), inverse=1 → backward (e^{+i2πkn/N})
    !> Returns 0 on success, -1 if N is not a power of two.
    function ufft_inplace(xr, xi, N, inverse) result(rc)
        integer,    intent(in)    :: N, inverse
        real(dp),   intent(inout) :: xr(0:N-1), xi(0:N-1)
        integer :: rc, len, i, j
        real(dp) :: ang, wr, wi, wre, wim, tre, tim, wre_new
        ! check power of two
        if (N <= 0 .or. iand(N, N-1) /= 0) then
            rc = -1; return
        end if
        call bit_reverse(xr, xi, N)
        len = 2
        do while (len <= N)
            ang = merge(-1.0_dp, 1.0_dp, inverse == 0) * 2.0_dp * PI / real(len, dp)
            wr = cos(ang); wi = sin(ang)
            i = 0
            do while (i < N)
                wre = 1.0_dp; wim = 0.0_dp
                do j = 0, len/2 - 1
                    tre = wre * xr(i+j+len/2) - wim * xi(i+j+len/2)
                    tim = wre * xi(i+j+len/2) + wim * xr(i+j+len/2)
                    xr(i+j+len/2) = xr(i+j) - tre
                    xi(i+j+len/2) = xi(i+j) - tim
                    xr(i+j) = xr(i+j) + tre
                    xi(i+j) = xi(i+j) + tim
                    wre_new = wre*wr - wim*wi
                    wim     = wre*wi + wim*wr
                    wre     = wre_new
                end do
                i = i + len
            end do
            len = len * 2
        end do
        rc = 0
    end function

    ! ── CC forward ────────────────────────────────────────────────────────────
    function ufft_cc_forward(in_r, in_i, out_r, out_i, N, dt) result(rc)
        integer,  intent(in)  :: N
        real(dp), intent(in)  :: in_r(0:N-1), in_i(0:N-1), dt
        real(dp), intent(out) :: out_r(0:N-1), out_i(0:N-1)
        integer :: rc
        out_r = in_r; out_i = in_i
        rc = ufft_inplace(out_r, out_i, N, 0)
        if (rc == 0) then
            out_r = out_r * dt
            out_i = out_i * dt
        end if
    end function

    ! ── CC inverse ────────────────────────────────────────────────────────────
    function ufft_cc_inverse(in_r, in_i, out_r, out_i, N, dt) result(rc)
        integer,  intent(in)  :: N
        real(dp), intent(in)  :: in_r(0:N-1), in_i(0:N-1), dt
        real(dp), intent(out) :: out_r(0:N-1), out_i(0:N-1)
        integer :: rc
        real(dp) :: df
        df = 1.0_dp / (real(N, dp) * dt)
        out_r = in_r; out_i = in_i
        rc = ufft_inplace(out_r, out_i, N, 1)   ! inverse twiddles
        if (rc == 0) then
            out_r = out_r * df   ! raw sum * Δf  (ufft_inplace has no 1/N)
            out_i = out_i * df
        end if
    end function

    ! ── CD forward ────────────────────────────────────────────────────────────
    function ufft_cd_forward(in_r, in_i, out_r, out_i, N) result(rc)
        integer,  intent(in)  :: N
        real(dp), intent(in)  :: in_r(0:N-1), in_i(0:N-1)
        real(dp), intent(out) :: out_r(0:N-1), out_i(0:N-1)
        integer :: rc
        out_r = in_r; out_i = in_i
        rc = ufft_inplace(out_r, out_i, N, 0)
        if (rc == 0) then
            out_r = out_r / real(N, dp)
            out_i = out_i / real(N, dp)
        end if
    end function

    ! ── CD inverse ────────────────────────────────────────────────────────────
    function ufft_cd_inverse(in_r, in_i, out_r, out_i, N) result(rc)
        integer,  intent(in)  :: N
        real(dp), intent(in)  :: in_r(0:N-1), in_i(0:N-1)
        real(dp), intent(out) :: out_r(0:N-1), out_i(0:N-1)
        integer :: rc
        out_r = in_r; out_i = in_i
        rc = ufft_inplace(out_r, out_i, N, 1)   ! raw sum — no extra factor
    end function

    ! ── Frequency bin array ──────────────────────────────────────────────────
    subroutine ufft_freqs(f, N, dt)
        integer,  intent(in)  :: N
        real(dp), intent(in)  :: dt
        real(dp), intent(out) :: f(0:N-1)
        real(dp) :: df
        integer  :: k
        df = 1.0_dp / (real(N, dp) * dt)
        do k = 0, N-1
            if (k < N/2) then
                f(k) = real(k, dp) * df
            else
                f(k) = real(k - N, dp) * df
            end if
        end do
    end subroutine

    ! ── Low-pass brick-wall response ─────────────────────────────────────────
    subroutine ufft_low_pass(H, f, N, fc)
        integer,  intent(in)  :: N
        real(dp), intent(in)  :: f(0:N-1), fc
        real(dp), intent(out) :: H(0:N-1)
        integer :: k
        do k = 0, N-1
            H(k) = merge(1.0_dp, 0.0_dp, abs(f(k)) <= fc)
        end do
    end subroutine

    ! ── Filter via CC pair ────────────────────────────────────────────────────
    function ufft_filter(in_r, in_i, H_r, H_i, out_r, out_i, N, dt) result(rc)
        integer,  intent(in)  :: N
        real(dp), intent(in)  :: in_r(0:N-1), in_i(0:N-1)
        real(dp), intent(in)  :: H_r(0:N-1),  H_i(0:N-1)
        real(dp), intent(in)  :: dt
        real(dp), intent(out) :: out_r(0:N-1), out_i(0:N-1)
        integer :: rc, k
        real(dp) :: Xr(0:N-1), Xi(0:N-1), yr, yi
        rc = ufft_cc_forward(in_r, in_i, Xr, Xi, N, dt)
        if (rc /= 0) return
        do k = 0, N-1
            yr = Xr(k)*H_r(k) - Xi(k)*H_i(k)
            yi = Xr(k)*H_i(k) + Xi(k)*H_r(k)
            Xr(k) = yr; Xi(k) = yi
        end do
        rc = ufft_cc_inverse(Xr, Xi, out_r, out_i, N, dt)
    end function

end module universalfft_mod
